extensions/scarb-snforge-test-collector/src/compilation/test_collector.rs

```diff
@@ -0,0 +1,244 @@
+use cairo_lang_semantic::items::functions::GenericFunctionId;
+use cairo_lang_semantic::{ConcreteFunction, FunctionLongId};
+use cairo_lang_sierra::extensions::enm::EnumType;
+use cairo_lang_sierra::extensions::NamedType;
+use cairo_lang_sierra::program::{GenericArg, Program};
+use cairo_lang_sierra_generator::db::SierraGenGroup;
+use cairo_lang_sierra_generator::replace_ids::replace_sierra_ids_in_program;
+use cairo_lang_starknet::starknet_plugin_suite;
+use cairo_lang_test_plugin::test_config::TestExpectation;
+use cairo_lang_test_plugin::test_plugin_suite;
+use cairo_lang_utils::ordered_hash_map::OrderedHashMap;
+use itertools::Itertools;
+use serde::Serialize;
+use smol_str::SmolStr;
+use std::path::{Path, PathBuf};
+use std::sync::Arc;
+
+use crate::metadata::LinkedLibrary;
+use config::{forge_try_extract_test_config, FuzzerConfig, SingleTestConfig};
+pub use config::{ForkConfig, RawForkConfig};
+use plugin::snforge_test_plugin_suite;
+use sierra_casm_generator::SierraCasmGenerator;
+
+mod config;
+mod plugin;
+mod sierra_casm_generator;
+
+/// Finds the tests in the requested crates.
+pub fn find_all_tests(
+    db: &dyn SemanticGroup,
+    main_crate: CrateId,
+) -> Vec<(FreeFunctionId, SingleTestConfig)> {
+    let mut tests = vec![];
+    let modules = db.crate_modules(main_crate);
+    for module_id in modules.iter() {
+        let Ok(module_items) = db.module_items(*module_id) else {
+            continue;
+        };
+        tests.extend(module_items.iter().filter_map(|item| {
+            let ModuleItemId::FreeFunction(func_id) = item else {
+                return None;
+            };
+            let Ok(attrs) = db.function_with_body_attributes(FunctionWithBodyId::Free(*func_id))
+            else {
+                return None;
+            };
+            Some((
+                *func_id,
+                forge_try_extract_test_config(db.upcast(), &attrs).unwrap()?,
+            ))
+        }));
+    }
+    tests
+}
+
+#[derive(Debug, PartialEq, Clone, Serialize)]
+pub struct TestCase<T: ForkConfig> {
+    pub name: String,
+    pub available_gas: Option<usize>,
+    pub ignored: bool,
+    pub expected_result: TestExpectation,
+    pub fork_config: Option<T>,
+    pub fuzzer_config: Option<FuzzerConfig>,
+}
+
+pub type TestCaseRaw = TestCase<RawForkConfig>;
+
+pub fn collect_tests(
+    crate_name: &str,
+    crate_root: &Path,
+    lib_content: &str,
+    linked_libraries: &[LinkedLibrary],
+    corelib_path: PathBuf,
+) -> Result<(Program, Vec<TestCaseRaw>)> {
+    let crate_roots: OrderedHashMap<SmolStr, PathBuf> = linked_libraries
+        .iter()
+        .cloned()
+        .map(|source_root| (source_root.name.into(), source_root.path))
+        .collect();
+
+    let project_config = ProjectConfig {
+        base_path: crate_root.into(),
+        corelib: Some(Directory::Real(corelib_path)),
+        content: ProjectConfigContent {
+            crate_roots,
+            crates_config: Default::default(),
```

Comment:
Crate editions should be read from scarb metadata (available since 1.9.0 https://docs.rs/scarb-metadata/latest/scarb_metadata/struct.PackageMetadata.html#structfield.edition) and passed here as in https://github.com/software-mansion/scarb/blob/fcc3307b4b3f59b7abb16e07cf4048dea780e371/scarb/src/compiler/db.rs#L94